### ISSUE TITLE
Support application matcher for wlroots-based compositors

### ISSUE THREAD

**Issue by derzahla** [k0kubun/xremap#141]
Hi, after spending a good deal of time trying to debug what was wrong, I have not made a lot of progress.  I installed the xremap-x11-bin packing on Arch from the AUR sinceI read that wayland was also supported.  It seemed to work great at first but as my xremap.yaml file got more and more complex I began to notice shortcuts breaking that I was certain had just worked.   As I was continually adding new shortcuts though I figured it was something I had done somehow.  When I eliminated that possibility by keeping my config static and still noticed irregularies I went looking on github and realized there were 4 different versions being distributed. xremap for x11,  xremap for gnome on wayland, xremap for sway and xremap for "others".  After reading through all the issues and trying each variation of xremap I'm still confused.  

It would seem that the reason for 4 different versions is to support application match, though I didnt see that explicitly stated anywhere.    I know application matching was working  for me, at least sometimes, with Brave-browser-dev when using the xremap-x11-bin Arch package - even though I am running wayland and brave should run as a native app(not using Xwayland)

Is the "others" version supposed not support any application matching at all?

I thought I had some application matching success with the gnome and/or sway version but I might be imagining things and I did remove the "Application:" section at times while testing.  Adding further to my confusion, the xremap version I built from source for "others" using no "--features" flags at all, shows output indicating it is expecting sway to be running:

Failed to locate a SWAYSOCK from /run/user/1000/sway-ipc.*

Is there any hope for me getting dependable application matching with my setup or should I give up for the time being?

**Comment by k0kubun** [comment]
It seems like you're using Wayland. Then please avoid xremap-x11-bin (the package is not something I maintain, so I'm not sure about the "I read that wayland was also supported" part). I'm still not sure from your description whether you use Sway or GNOME. Please build xremap with `--features sway` if you use Sway, and use `--features gnome` if you use GNOME. If you use something else, please let me know.

If you're a GNOME Wayland user, you need a GNOME shell extension as instructed at https://github.com/k0kubun/xremap#gnome-wayland. Sway doesn't need a Sway-specific setup.

Lastly, using a root user sometimes causes problems in some environments, so you may want to try starting xremap without sudo if you were starting xremap with sudo. After clicking "If you want to run xremap without sudo, click here" under https://github.com/k0kubun/xremap#usage, please follow the description.

**Comment by derzahla** [comment]
the xremap-x11-bin PKGBUILD just downloads your package and installs it into the Arch package system.  I am using wayland  and my window manager is wayfire. I am not using gnome or sway.  I also have the permissions set and running xremap as my user, not root.

**Comment by k0kubun** [comment]
Ah, got it. I'm sorry that I overlooked that from the issue subject.

Long story short, the current version of xremap doesn't support matching applications on Wayfire. Wayland is still so experimental that the Wayland protocol does not support such a capability yet. Therefore, for Wayland compositors, each platform needs to be supported one by one, GNOME and Sway being the existing examples. You became the first person in this repository to request Wayfire support in xremap.

The good news is that Wayfire seems like a wlroots-based compositor. I've been planning to rewrite the Sway support to use the wlroots protocol so that it can also support other wlroots-based compositors such as Wayfire, while it's currently using a Sway-specific protocol. Once it's done, you should be able to match applications. I'm planning to spend some time fixing xremap issues this weekend, so stay tuned.

**Comment by derzahla** [comment]
awesome! that would be great.  Let me know if you would like me to test something. in the meantime, the x11 version has been working mostly for some apps (the ones that are x11 compatable i guess?)  like chromium based browsers.  Seems like running with "--watch" was causing issues for me as well so I stopped that, made a systemd service for xremap, and have been restarting it manually when i make changes.

**Comment by k0kubun** [comment]
I ended up burning out my entire weekend to work on other issues in xremap https://github.com/k0kubun/xremap/compare/f34e0811338530f04f6c786bbfa5e25056858674...e8d30a3e38b3e686cfbabfd66760ddee7b95a95f. So I didn't get to work on wlroots support this time.

Meanwhile, please continue to use the X11 version and configure stuff based on what's working under XWayland. As I need to focus on conference preparations for a few weeks, please allow me to work on this after Sep 10th or so.

**Comment by derzahla** [comment]
Ok no worries.  I switched to a x11 terminal since it seems to work alright with native X11 apps. Biggest problem is that after a a day or so it stops working but xremap is still running.  So I have to restart it to get it working again. 

Would be great to have wayland apps under wayfire work though. I wonder if its something simple enough that i could play around with the code and get it working.  How would pulling the focused window properties under differ wayfire vs other wayland WMs?

**Comment by k0kubun** [comment]
> Biggest problem is that after a a day or so it stops working but xremap is still running. So I have to restart it to get it working again.

If you want me to fix it, you need to provide steps to reproduce the problem. I haven't faced any of such issues in recent versions, so detailed instructions (but the minimum effort required for reproducing it at the same time) would be appreciated.

> I wonder if its something simple enough that i could play around with the code and get it working. How would pulling the focused window properties under differ wayfire vs other wayland WMs?

The plan is to replace the code that is currently Sway-specific with a wlroots one. I thought reading https://github.com/ActivityWatch/aw-watcher-window-wayland might be helpful.

**Comment by derzahla** [comment]
Sorry, I just saw your reply. So the main issue should be easily reproducable

1. Run wayland and configure wayfire as your window manager
2. Attempt to remap any keys for any native wayland application.

It will not work any flavor of xremap (x11,gnome,sway,  or other).  However, xremap-x11 will work somewhat reliably  with x11 apps running with xwayland.

I have been using the wf-prop utility included with [this project](https://github.com/damianatorrpm/wayfire-plugin_dbus_interface) to grab the properties of any running window to determine name, whether it is native or x11, etc.

**Comment by k0kubun** [comment]
The project seems to use wayfire-specific protocol, so it'd be to add a fourth flavor to the xremap binary distribution. It works for me too, but note that the plan is to eventually switch to wlroots protocl when applicable, so that will be removed at that point.

**Comment by RedBearAK** [comment]
@k0kubun 

Is there actually any sign that it might be possible to get the focused window attributes (class and name/title) directly from some sort of interaction with `wlroots`? What's you're plan on this? 

I just got done adding support for `sway` to my development branch of `keyszer`, using the `i3ipc` Python package. Sure would be nice if there was a more widely compatible way to deal with some of the Wayland window managers.

**Comment by k0kubun** [comment]
I haven't investigated that yet. I can share my experience once I'm done, but I'm not yet. If you get to work on that earlier than me, I'd appreciate your input too.

**Comment by RedBearAK** [comment]
> I'd appreciate your input too.

I will try to come back to this and leave a note if I ever find anything. My AI assistant seems pretty confident that the necessary methods such as IPC sockets are probably only implemented by the actual compositors that use `wlroots` as a base. It still has a training cut-off date of about 2 years ago, but `wlroots` has been around for several years. I have a feeling our use case is simply not something that the `wlroots` devs have designed into their components. 

We will see what happens as the Wayland transition progresses. Fedora is talking about dropping X11/Xorg in the F40 release that would come out in only about 9 months, to spur additional progress toward a complete transition. It's still just in discussion, but I won't be surprised if they do it, to follow along with Red Hat's plans for dropping X11/Xorg from RHEL 10. At most they may push it 6 months to the F41 release.

**Comment by meowcoder** [comment]
I've implemented support for `wlroots` here  43eb7fbdec5a466c8c51e6ac3cb9a5420c5ac05d. 
Only app_id / class is tracked, no window titles, although it's possible to track them as well. 
I've been using it in `labwc` for awhile and now `sway`, haven't noticed any problems.

**Comment by RedBearAK** [comment]
@meowcoder 

> I've implemented support for wlroots here

This is a nice thing to see in my notifications. 

I'm taking a look. Do you have any idea if the method you used could be accomplished in Python? I always have to use AI to help me analyze these Rust modules and re-implement the methods using Python. 

>  Only app_id / class is tracked, no window titles, although it's possible to track them as well.

I absolutely need the window name, unfortunately, so in addition to translating this into something that works in Python, I'll need to add that. 

Thanks for posting this!

**Comment by meowcoder** [comment]
> Do you have any idea if the method you used could be accomplished in Python?

Seems like pywayland doesn't support one of the wayland protocols I used:
flacjacket/pywayland#33 
You would have to fix the issue first. 
Or find alternative bindings, maybe make your own. 

> I absolutely need the window name, unfortunately, so in addition to translating this into something that works in Python, I'll need to add that.

It's not difficult to add, just an extra match case and a dict to store the titles.

**Comment by RedBearAK** [comment]
@meowcoder 

It became pretty obvious that something was missing as soon as I tried to use the same protocol. I grabbed the XML for the protocol: 

https://gitlab.freedesktop.org/wlroots/wlr-protocols/-/blob/master/unstable/wlr-foreign-toplevel-management-unstable-v1.xml  

...  and generated the Python modules with `pywayland.scanner`:

```sh
python -m pywayland.scanner -i wayland.xml wlr-foreign-toplevel-management-unstable-v1.xml -o .
```

Then placed everything it generated into the same folder as the Python module that needs to use it. 

Trying to work my way through figuring out how to use it. We'll see how it goes.

**Comment by k0kubun** [comment]
> I've implemented support for wlroots here https://github.com/k0kubun/xremap/commit/43eb7fbdec5a466c8c51e6ac3cb9a5420c5ac05d.

:+1: Could you make it a pull request?

> Only app_id / class is tracked, no window titles, although it's possible to track them as well.

We want to add window title support, but I think we currently don't have that, do we? We should be able to merge yours as is for now, and then work on improving it later.

**Comment by k0kubun** [comment]
Released Alex's contribution as v0.8.9. Published a package for it on AUR https://aur.archlinux.org/packages/xremap-wlroots-bin.
